torch/csrc/autograd/record_function.cpp

```diff
@@ -242,6 +242,53 @@ RecordFunction* RecordFunction::current() {
   return thread_local_func_;
 }
 
+void RecordFunctionAsync::before(const char* name, int64_t sequence_nr) {
+  RecordFunction::before(name, sequence_nr);
+  RecordFunction::setThreadId();
+}
+
+void RecordFunctionAsync::before(std::string name, int64_t sequence_nr) {
+  RecordFunction::before(name, sequence_nr);
+  RecordFunction::setThreadId();
+}
+
+void RecordFunctionAsync::before(Node* fn, int64_t sequence_nr) {
+  RecordFunction::before(fn, sequence_nr);
+   RecordFunction::setThreadId();
+}
+
+void RecordFunctionAsync::exitScope() {
+  // We should not be calling exitScope() on an uninitialized RecordFunction.
+  TORCH_INTERNAL_ASSERT(
+      initialized_, "Current RecordFunction is not initialized.")
+  // If the current RecordFunction is not the stored thread_local
+  // RecordFunction, the scoping is in a bad state.
+  TORCH_INTERNAL_ASSERT(
+      thread_local_func_ == this, name_, ": must be top of stack.");
+  // Resets the thread_local func to the parent_ RecordFunction that outlives
+  // this RecordFunction, to correctly keep track of scopes.
+  thread_local_func_ = parent_;
+}
+
+void RecordFunctionAsync::end() {
+  if (initialized_) {
+    for (size_t idx = 0; idx < manager().end_callbacks.size(); ++idx) {
+      if (!manager().is_callback_sampled[idx] || run_sampled_) {
+        manager().end_callbacks[idx](*this);
+      }
+    }
+    initialized_ = false;
+  }
+}
+
+// toggle initialized so that RecordFunction destructor does not attempt to
+// reinvoke callbacks.
+RecordFunctionAsync::~RecordFunctionAsync() {
```

Comment:
If `exitScope` and `end` haven't been called yet, this should be an error here right? Can we add a `DCHECK` here ensuring both `exitScope` and `end` have been called?

Reply:
I guess it is a logic error, though technically, if the user just creates and destroys a `RecordFunctionAsync`, nothing will happen, since:
- before() is not called, so starting callbacks are not run
- end() is not called, so ending callbacks are not run, and they are not run in the RecordFn destructor either, since we reset initialized_ = false.

If we decide to check, is the best way to do this to add booleans indicating the functions have been called?

Reply:
Also `RecordFunction` doesn't currently contain error handling if `before()` is not called at all or called after `End`, maybe we should check this also if we decide to go this route.

Reply:
I added some error handling in the top diff in this stack.

Reply:
If thread_local_func == this, it means exit_scope hasn't been called, if initialized_ = true, it means end() hasn't been called. We can raise errors for these cases.

Reply:
> Also RecordFunction doesn't currently contain error handling if before() is not called at all or called after End, maybe we should check this also if we decide to go this route.

Might be fine for RecordFunction since its single threaded and the RECORD_FUNCTION guard takes care of most things.

Reply:
> If thread_local_func == this, it means exit_scope hasn't been called, if initialized_ = true, it means end() hasn't been called. We can raise errors for these cases.

Should we log them or actually call `TORCH_INTERNAL_ASSERT` that will throw an exception? This is a destructor and I thought it was generally bad to throw in the destructor. Also, I think clang-tidy would complain if we did.